File: frontend/src/api/campaign.ts
import client from './client'
import type { ListResponse } from './types'

export interface Campaign {
  id: number
  organization_id: number
  name: string
  description: string | null
  campaign_type: string
  status: string
  channels: string[] | null
  target_audience: Record<string, unknown> | null
  target_count: number
  budget: number
  actual_cost: number
  start_date: string | null
  end_date: string | null
  sent_count: number
  opened_count: number
  converted_count: number
  converted_revenue: number
  created_at: string
  updated_at: string
}

export interface CampaignCreate {
  name: string
  description?: string
  campaign_type?: string
  channels?: string[]
  target_audience?: Record<string, unknown>
  target_count?: number
  budget?: number
  start_date?: string
  end_date?: string
}

export interface CampaignUpdate {
  name?: string
  description?: string
  status?: string
  channels?: string[]
  target_audience?: Record<string, unknown>
  target_count?: number
  budget?: number
  actual_cost?: number
  start_date?: string
  end_date?: string
  sent_count?: number
  opened_count?: number
  converted_count?: number
  converted_revenue?: number
}

export const campaignApi = {
  getList: async (params?: { skip?: number; limit?: number; status?: string }): Promise<ListResponse<Campaign>> => {
    const response = await client.get('/api/v1/campaigns/', { params })
    return response.data
  },

  get: async (id: number): Promise<Campaign> => {
    const response = await client.get(`/api/v1/campaigns/${id}`)
    return response.data
  },

  create: async (data: CampaignCreate): Promise<Campaign> => {
    const response = await client.post('/api/v1/campaigns/', data)
    return response.data
  },

  update: async (id: number, data: CampaignUpdate): Promise<Campaign> => {
    const response = await client.put(`/api/v1/campaigns/${id}`, data)
    return response.data
  },

  delete: async (id: number): Promise<void> => {
    await client.delete(`/api/v1/campaigns/${id}`)
  },
}

File: backend/schemas/campaign.py
from datetime import datetime
from typing import Optional
from pydantic import BaseModel

from backend.models.campaign import CampaignStatus


class CampaignCreate(BaseModel):
    name: str
    description: Optional[str] = None
    campaign_type: str = "promotion"
    channels: Optional[list[str]] = None
    target_audience: Optional[dict] = None
    target_count: int = 0
    budget: float = 0
    start_date: Optional[datetime] = None
    end_date: Optional[datetime] = None


class CampaignUpdate(BaseModel):
    name: Optional[str] = None
    description: Optional[str] = None
    status: Optional[CampaignStatus] = None
    channels: Optional[list[str]] = None
    target_audience: Optional[dict] = None
    target_count: Optional[int] = None
    budget: Optional[float] = None
    actual_cost: Optional[float] = None
    start_date: Optional[datetime] = None
    end_date: Optional[datetime] = None
    sent_count: Optional[int] = None
    opened_count: Optional[int] = None
    converted_count: Optional[int] = None
    converted_revenue: Optional[float] = None


class CampaignResponse(BaseModel):
    id: int
    organization_id: int
    name: str
    description: Optional[str] = None
    campaign_type: str
    status: CampaignStatus
    channels: Optional[list] = None
    target_audience: Optional[dict] = None
    target_count: int
    budget: float
    actual_cost: float
    start_date: Optional[datetime] = None
    end_date: Optional[datetime] = None
    sent_count: int
    opened_count: int
    converted_count: int
    converted_revenue: float
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}
